aws/resource_aws_cognito_identity_provider.go

```diff
@@ -0,0 +1,181 @@
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsCognitoIdentityProvider() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCognitoIdentityProviderCreate,
+		Read:   resourceAwsCognitoIdentityProviderRead,
+		Update: resourceAwsCognitoIdentityProviderUpdate,
+		Delete: resourceAwsCognitoIdentityProviderDelete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"attribute_mapping": {
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+
+			"idp_identifiers": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"provider_details": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Computed: true,
+			},
+
+			"provider_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"provider_type": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"user_pool_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsCognitoIdentityProviderCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cognitoidpconn
+	log.Print("[DEBUG] Creating Cognito Identity Provider")
+
+	name := aws.String(d.Get("provider_name").(string))
+	params := &cognitoidentityprovider.CreateIdentityProviderInput{
+		ProviderName: name,
+		ProviderType: aws.String(d.Get("provider_type").(string)),
+		UserPoolId:   aws.String(d.Get("user_pool_id").(string)),
+	}
+
+	if v, ok := d.GetOk("attribute_mapping"); ok {
+		params.AttributeMapping = expandCognitoIdentityProviderMap(v.(map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("provider_details"); ok {
+		params.ProviderDetails = expandCognitoIdentityProviderMap(v.(map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("idp_identifiers"); ok {
+		params.IdpIdentifiers = expandStringList(v.([]interface{}))
+	}
+
+	_, err := conn.CreateIdentityProvider(params)
+	if err != nil {
+		return fmt.Errorf("Error creating Cognito Identity Provider: %s", err)
+	}
+
+	d.SetId(*name)
+
+	return resourceAwsCognitoIdentityProviderRead(d, meta)
+}
+
+func resourceAwsCognitoIdentityProviderRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cognitoidpconn
+	log.Printf("[DEBUG] Reading Cognito Identity Provider: %s", d.Id())
+
+	ret, err := conn.DescribeIdentityProvider(&cognitoidentityprovider.DescribeIdentityProviderInput{
+		ProviderName: aws.String(d.Id()),
+		UserPoolId:   aws.String(d.Get("user_pool_id").(string)),
+	})
+
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" {
```

Comment:
Minor nitpick: we have a helper function to simplify this logic: `isAWSErr(err, cognitoidentity.ErrCodeResourceNotFoundException, "")`